### PR TITLE
Fixed double click bug in quality control view

### PR DESCRIPTION
**Pull request by Centauri0320001**
### Prerequisites

- [x] Reviewed the [checklist](CHECKLIST.md)

- [ ] Reviewed feedback from the "Sonar Cloud" bot. Note that you have to wait
    for the "CI / Unit Tests") to complete first. Failed Unit tests can be 
    debugged by adding the label "verbose logging" to the GitHub PR.

### Description of the Change

Fixed bug where the users first click would count as two clicks in the Quality Control View when a new sphere graph is created. 
### Why Should This Be In Core?

Fixes bug which causes confusion when using quality control view.
### Benefits

Improves usability of quality control view.
### Applicable Issues

#784

### CODE REVIEW

**Review by OrionsGuardian — CHANGES_REQUESTED** [review]
The problem is not fully fixed.
Only the very first click in the Quality Control table is blocked.
If you then de-select then re-select the same node in the graph, and single click the Quality Control table entry, it will again be treated like a double click.

File: CoreQualityControlView/src/au/gov/asd/tac/constellation/views/qualitycontrol/QualityControlViewTopComponent.java [hunk 1]
```diff
@@ -53,6 +53,8 @@
     "HINT_QualityControlViewTopComponent=Quality Control View"})
 public final class QualityControlViewTopComponent extends JavaFxTopComponent<QualityControlViewPane> implements QualityControlListener {
 
+    private static boolean firstClick = true;
```
Comment (OrionsGuardian) [inline CoreQualityControlView/src/au/gov/asd/tac/constellation/views/qualitycontrol/QualityControlViewTopComponent.java:56]: This variable should be defined in the QualityControlViewPane class.
There's no need for any code changes in this class.

File: CoreQualityControlView/src/au/gov/asd/tac/constellation/views/qualitycontrol/QualityControlViewPane.java [hunk 2]
```diff
@@ -314,11 +317,13 @@ public void updateItem(final QualityControlEvent item, final boolean empty) {
                 };
 
                 cell.setOnMouseClicked(value -> {
-                    if (value.getClickCount() == 2) {
+                    if (value.getClickCount() == 2 && !QualityControlViewTopComponent.isFirstClick() || 
+                            value.getClickCount() > 2 && !QualityControlViewTopComponent.isFirstClick()) {
```
Comment (OrionsGuardian) [inline CoreQualityControlView/src/au/gov/asd/tac/constellation/views/qualitycontrol/QualityControlViewPane.java:321]: The `if` conditions can be simplified to:
```
        if (value.getClickCount() >= 2 && !firstClick) {
```
And have `firstClick` defined as a local class member variable.

**Review by OrionsGuardian — CHANGES_REQUESTED** [review]

**Review by Delphinus8821 — CHANGES_REQUESTED** [review]
I also found that if you click off of Constellation and then single click back onto the quality control view it is treated like a double click.

**Review by OrionsGuardian — CHANGES_REQUESTED** [review]
Also need to consider the case where a user may genuinely double-click on an entry in the table as the first thing they do.

File: CoreQualityControlView/src/au/gov/asd/tac/constellation/views/qualitycontrol/QualityControlViewTopComponent.java [hunk 3]
```diff
@@ -89,6 +90,17 @@ public void handleComponentClosed() {
...
+    @Override 
+    public void handleNewGraph(final Graph graph) {
+        qualityControlViewPane.setFirstClick(true); 
+    }
+    
+    @Override 
+    protected void handleGraphChange(final GraphChangeEvent event) {
+        qualityControlViewPane.setFirstClick(true);
+    }
+
```
Comment (OrionsGuardian) [inline CoreQualityControlView/src/au/gov/asd/tac/constellation/views/qualitycontrol/QualityControlViewTopComponent.java:103]: I don't believe these changes are required.
The focus listener in the View Pane ensures the firstClick variable will always be set to true whenever the View Pane gets focus.

File: CoreQualityControlView/src/au/gov/asd/tac/constellation/views/qualitycontrol/QualityControlViewPane.java [hunk 4]
```diff
@@ -683,6 +692,11 @@ private static void showRuleDialog(final String identifier, final List<Pair<Qual
         alert.setHeaderText(String.format(Bundle.MSG_QualtyControlRules(), identifier));
         alert.getDialogPane().setContent(sp);
         alert.setResizable(true);
+        final List<Screen> screens = Screen.getScreensForRectangle(this.getScene().getWindow().getX(), this.getScene().getWindow().getY(),
+                this.getScene().getWindow().widthProperty().get(), this.getScene().getWindow().heightProperty().get());
+
+        alert.setX((screens.get(0).getVisualBounds().getMinX() + screens.get(0).getVisualBounds().getWidth() / 2));
+        alert.setY((screens.get(0).getVisualBounds().getMinY() + screens.get(0).getVisualBounds().getHeight() / 2));
```
Comment (OrionsGuardian) [inline CoreQualityControlView/src/au/gov/asd/tac/constellation/views/qualitycontrol/QualityControlViewPane.java:699]: This needs to account for the size of the Rule Dialog as well.
As it is, it shows in the bottom right portion of the application screen, with the bottom of the dialog potentially going below the bottom of the screen.

File: CoreQualityControlView/src/au/gov/asd/tac/constellation/views/qualitycontrol/QualityControlViewPane.java [hunk 5]
```diff
@@ -360,6 +367,10 @@ public void updateItem(final QualityControlEvent item, final boolean empty) {
         });
     }
 
+    public void setFirstClick(final boolean click) {
+        firstClick = click;
+    }
```
Comment (OrionsGuardian) [inline CoreQualityControlView/src/au/gov/asd/tac/constellation/views/qualitycontrol/QualityControlViewPane.java:372]: Do not need a publicly accessible setter for this variable.
Assign the values directly.

**Review by OrionsGuardian — CHANGES_REQUESTED** [review]

File: CoreQualityControlView/src/au/gov/asd/tac/constellation/views/qualitycontrol/QualityControlViewTopComponent.java [hunk 6]
```diff
@@ -16,6 +16,7 @@
 package au.gov.asd.tac.constellation.views.qualitycontrol;
 
 import au.gov.asd.tac.constellation.graph.Graph;
+import au.gov.asd.tac.constellation.graph.monitor.GraphChangeEvent;
```
Comment (OrionsGuardian) [inline CoreQualityControlView/src/au/gov/asd/tac/constellation/views/qualitycontrol/QualityControlViewTopComponent.java:19]: remove unused import

**Review by OrionsGuardian — CHANGES_REQUESTED** [review]

File: CoreQualityControlView/src/au/gov/asd/tac/constellation/views/qualitycontrol/QualityControlViewTopComponent.java
Comment (OrionsGuardian) [inline CoreQualityControlView/src/au/gov/asd/tac/constellation/views/qualitycontrol/QualityControlViewTopComponent.java:1]: Given that the only changes are now whitespace changes, I'd suggest undoing the changes. 
No value in checking in an essentially unchanged file.

**Review by OrionsGuardian — CHANGES_REQUESTED** [review]

**Review by OrionsGuardian — APPROVED** [review]
The code change does not identify the underlying cause of the double-click problem,
however, it does adequately address the functionality issue.
It is an acceptable workaround.

**Review by mimosa2 — APPROVED** [review]

**Review by antares1470 — CHANGES_REQUESTED** [review]
Agreeing with @OrionsGuardian in that this will be an adequate workaround since the actual issue (with the ControlsFX object) can't be fixed within Constellation itself.

With that in mind, are you possibly able to add some comments just to note that it is a workaround (possibly add it at the declaration of `firstClick`)? Doing so will just help us remember to re-assess the underlying issue in the future (as hopefully if the fx object is fixed, this workaround shouldn't be necessary)

**Review by antares1470 — APPROVED** [review]

**Pull request merged**